File: 02-week-2--recursion/05-day-5--recursive-selection-sort/ruby/selection_sort_recursive.rb
def selection_sort_recursive(arr)
  # type your code in here
  if arr.length == 0 || arr.length == 1
    arr
  else
    max = arr.max
    index = arr.index(max)
    arr.slice!(index)
    new_arr = selection_sort_recursive(arr)
    new_arr.push(max)
  end

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort_recursive([3, -1, 5, 2])

  puts

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
=begin
  -if the length of the argument array is equal to 0 or 1 return the argument array
  -else assign the highest element in the argument array *arr to a variable *max,
  -find the index of max within arr and assign it to a variable *index
  -remove that index from arr
  -assign a variable *new_arr to the return of this function using the now shortened arr as the argument
  -return new_arr with the previously assignmed maximum value variable max at the end.
=end
# And a written explanation of your solution
=begin
  If the arrays length was either 0 or 1 I had to immediately return that array.  If that wasn't the case, with each recursion I
  needed to find the maximum value and assign it to a variable, remove that maximum value from the array, assign a new variable
  using a recursion of this function with the now shortened original argument array, and return this new array variable with the
  previously defined maximum element at the end.
=end

File: 02-week-2--recursion/05-day-5--recursive-selection-sort/javascript/selection_sort_recursive.js
function selectionSortRecursive(arr) {
  // type your code here
  if (arr.length === 0 || arr.length === 1) {
    return arr;
  }

  const min = Math.min(...arr)
  const i = arr.indexOf(min);
  arr.splice(i, 1)

  const result = selectionSortRecursive(arr);
  result.unshift(min);
  return result
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSortRecursive([3, -1, 5, 2]));

  console.log("");
}

module.exports = selectionSortRecursive;

// Please add your pseudocode to this file
/*
  -if the length of the argument array *arr is either 0 or 1 return arr
  -find the smallest value in arr, store it in a variable called *min
  -find the index of min within arr, store it in a variable called *i
  -remove min from arr
  -declare a variable *result which is the recursive return of this function being passed arr.unshift(min)
*/
// And a written explanation of your solution
/*
  with each recursion I am finding the lowest element within the array, assigning it to a variable and removing it
  I am then passing the shortened array to this function for recursion and assigning the result to a variable
  I am then returning this variable with the original minimum value placed at the begining of it.
*/